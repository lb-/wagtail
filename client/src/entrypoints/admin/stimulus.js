import { Application, defaultSchema } from '@hotwired/stimulus';

/**
 *
 */
const SCHEMA = { ...defaultSchema };

/**
 * Extending the built in class for housing additional logic.
 *
 */
class WagtailApplication extends Application {
  // ...
}

const Stimulus = WagtailApplication.start(document.documentElement, SCHEMA);

Stimulus.debug = process.env.NODE_ENV === 'development';

const initEvent = new CustomEvent('wagtail:stimulus-init', {
  bubbles: true,
  cancelable: false,
  detail: {
    register: (identifier, controller) => {
      Stimulus.register(identifier, controller);
      Stimulus.logDebugActivity('registered', identifier);
    },
  },
});

// dispatch event for any core JS already loaded
document.dispatchEvent(initEvent);

// dispatch event for any other JS loaded async
document.addEventListener(
  'DOMContentLoaded',
  () => {
    document.dispatchEvent(initEvent);
  },
  { once: true },
);

/**
Example usage.

```python
@hooks.register('insert_global_admin_js')
def global_admin_js():
    return mark_safe(
        """
<script type="module">
import { Controller } from "https://unpkg.com/@hotwired/stimulus/dist/stimulus.js";

class HelloController extends Controller {
  static targets = ["name"];

  connect() {
    console.log("connected");
  }
}

document.addEventListener(
  "wagtail:stimulus-init",
  ({ detail: { register } }) => {
    register("hello", HelloController);
  },
  { once: true }
);
</script>
        """
    )
```
 */
