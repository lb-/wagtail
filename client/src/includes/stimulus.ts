import { Application, defaultSchema } from '@hotwired/stimulus';

import controllerDefinitions from '../controllers';

/**
 * Extending the built in class for housing additional logic.
 *
 * @example - within Wagtail for custom stimulus controllers
 * ```python
 * @hooks.register('insert_global_admin_js')
 * def global_admin_js():
 *     return mark_safe(
 *         """
 * <script type="module">
 * import { Controller } from "https://unpkg.com/@hotwired/stimulus/dist/stimulus.js";
 *
 * class HelloController extends Controller {
 *   static targets = ["name"];
 *
 *   connect() {
 *     console.log("connected");
 *   }
 * }
 *
 * document.addEventListener(
 *   "wagtail:stimulus-init",
 *   ({ detail: { register } }) => {
 *     register(["hello", HelloController]);
 *   },
 *   { once: true }
 * );
 * </script>
 *         """
 *     )
 * ```
 *
 */
class WagtailApplication extends Application {
  /**
   * Prepare for a custom schema as required.
   */
  static SCHEMA = { ...defaultSchema };
}

const initStimulus = () => {
  const application = WagtailApplication.start(
    document.documentElement,
    WagtailApplication.SCHEMA,
  );

  application.debug = process.env.NODE_ENV === 'development';

  application.load(controllerDefinitions);

  const getInitEvent = (order: number): CustomEvent =>
    new CustomEvent('wagtail:stimulus-init', {
      bubbles: true,
      cancelable: false,
      detail: {
        // intentionally not providing the application - may add this in a future release if needed
        order,
        // provide a way to register controllers
        register: ([identifier, controller]) => {
          application.register(identifier, controller);
          application.logDebugActivity('registered', identifier, { order });
        },
      },
    });

  // dispatch event for any core JS already loaded
  document.addEventListener('readystatechange', () =>
    document.dispatchEvent(
      getInitEvent(document.readyState === 'interactive' ? 0 : 2),
    ),
  );

  // dispatch event for any other JS loaded async
  document.addEventListener(
    'DOMContentLoaded',
    () => document.dispatchEvent(getInitEvent(1)),
    { once: true },
  );
};

export default initStimulus;
