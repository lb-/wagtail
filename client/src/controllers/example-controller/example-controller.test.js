import { Application } from '@hotwired/stimulus';

import ExampleController from './example-controller';

describe('ExampleController', () => {
  beforeAll(() => {
    const application = Application.start();
    application.register('example', ExampleController);

    document.body.innerHTML = '<h1 data-controller="example" id="example" data-example-name-value="Mr. Rabbit"></h1>';
  });

  it('should exist', () => {
    expect(ExampleController).toBeInstanceOf(Function);
  });

  it('should update the content of the controlled element', () => {
    expect(document.getElementById('example').textContent).toEqual('Hello, Mr. Rabbit!');
    expect(document.getElementById('example')).toMatchSnapshot();
  });
});
