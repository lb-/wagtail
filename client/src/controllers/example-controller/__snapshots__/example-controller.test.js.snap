// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleController should update the content of the controlled element 1`] = `
<h1
  data-controller="example"
  data-example-name-value="Mr. Rabbit"
  id="example"
>
  Hello, Mr. Rabbit!
</h1>
`;
